Improvements for recycler script

hansmi opened the pull request:
The `recycler.sh` script contained a number of issues preventing its use on production systems (including `eval` being used with potentially untrusted, improperly quoted data). See the descriptions of the individual commits for details.

dtschan commented:
Found two issues with the new code:

* Prints `Attempting to re-cycle volume` for volumes annotated for recycling, but not yet due, on each check.
* Sometimes crashes when trying to recycle a volume with:

```
/recycler.sh: line 371: /tmp/tmp.YkIkDoDB2z/failed.json: No such file or directory
/recycler.sh: line 374: /tmp/tmp.YkIkDoDB2z/failed.json: No such file or directory
2017-09-12T20:31:45+0200:  failed volumes found
Finished recycle run
```

But wasn't able to reproduce it reliably.

hansmi commented:
> Prints `Attempting to re-cycle volume` for volumes annotated for recycling, but not yet due, on each check.

Changed the code to emit `Annotating ${vol_name} as failed at ${failed_at}` before doing so and `Recycling volume ${vol_name}` when recycling, and nothing when the delay hasn't passed yet.

> Sometimes crashes when trying to recycle a volume with: `[…]/failed.json: No such file or directory[…]`

It appears as if something deletes the temporary directory. The script uses an `EXIT` trap to remove the temporary directory upon exit. I assume you ran the script in a container, which wouldn't have `tmpreaper(8)` or an equivalent utility removing temporary files/directories. While we could use a hardcoded path, I'd rather avoid that and find the cause. In which environment (base image version, etc.), and using which Bash version (`bash --version`) did you test?

dtschan commented:
I tested by deploying the recycler using the Ansible role in this repo on an OCP 3.4 cluster running on Vagrant, i.e. centos:7 and GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu).
I can now reproduce the problem. Delete or redeploy the pod. The scripts does not terminate when it reiceives the SIGTERM signal but continuous to run until the graceful shutdown period ends and it gets killed with SIGKILL. When a recycle run happens during the grace period the mentioned error gets logged. So the issue only happens on pod shutdown. Since I tried various interval and delay values I encountered several of those during testing.

hansmi commented:
> The scripts does not terminate when it reiceives the SIGTERM signal but continuous to run until the graceful shutdown period ends and it gets killed with SIGKILL

This is to be expected because a plain shell script used as an init process (PID 1) doesn't handle signals as required by the kernel. The only thing changed from before in this regard is that now a temporary directory is used and the effects are more visible.

Should we add [dumb-init](https://github.com/Yelp/dumb-init) to implement the signal handling required from PID 1?

tobru commented:
Yes, please use dumb-init or [tini](https://github.com/krallin/tini), whatever you like more.

hansmi commented:
> Yes, please use dumb-init or tini, whatever you like more.

Added tini (no hard preference between that and dumb-init), and added a patch to upgrade the Gluster client to version 3.12 (see commit message).

@dtschan, could you please test again in your environment?

dtschan commented:
RHEL based builds are now failing with:

```
Step 2 : RUN yum install -y --setopt=tsflags=nodocs centos-release-gluster312 &&   INSTALL_PKGS="bash tar jq findutils which glusterfs-fuse" &&   rpm -ihv https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm &&   yum install -y --setopt=tsflags=nodocs $INSTALL_PKGS &&   rpm -V $INSTALL_PKGS &&   yum clean all
 ---> Running in 7ac67fe6709b
Loaded plugins: ovl, product-id, search-disabled-repos, subscription-manager
No package centos-release-gluster312 available.
```

You'll probably have to use `rpm -i http://mirror.centos.org/centos/7/extras/x86_64/Packages/centos-release-gluster312-1.0-1.el7.centos.noarch.rpm` analogous to `epel-release` to make it work with RHEL.

Testing the other changes with CentOS base image now.

hansmi commented:
> RHEL based builds are now failing with: […]

Updated the code to install the necessary packages directly. Please note that I wasn't able to actually test with a RHEL base image, but I changed the condition to use the RHEL-specific code on CentOS for testing.
